File: app/src/main/java/igrek/todotree/ui/contextmenu/BackupListMenu.java
package igrek.todotree.ui.contextmenu;


import android.app.Activity;
import android.app.AlertDialog;
import android.content.DialogInterface;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import javax.inject.Inject;

import igrek.todotree.commands.GUICommand;
import igrek.todotree.commands.PersistenceCommand;
import igrek.todotree.dagger.DaggerIOC;
import igrek.todotree.service.backup.Backup;
import igrek.todotree.service.backup.BackupManager;
import igrek.todotree.service.resources.UserInfoService;
import igrek.todotree.service.tree.TreeManager;
import igrek.todotree.service.tree.TreeScrollCache;
import igrek.todotree.ui.errorcheck.UIErrorHandler;

public class BackupListMenu {
	
	@Inject
	Activity activity;
	
	@Inject
	BackupManager backupManager;
	
	@Inject
	UserInfoService userInfo;
	
	@Inject
	TreeScrollCache scrollCache;
	
	@Inject
	TreeManager treeManager;
	
	private SimpleDateFormat displayDateFormat = new SimpleDateFormat("yyyy-MM-dd, HH:mm:ss", Locale.ENGLISH);
	
	public BackupListMenu() {
		DaggerIOC.getFactoryComponent().inject(this);
	}
	
	public void show() {
		
		final List<RestoreBackupAction> actions = buildActionsList();
		CharSequence[] actionNames = convertToNamesArray(actions);
		
		AlertDialog.Builder builder = new AlertDialog.Builder(activity);
		builder.setTitle("Choose backup");
		builder.setItems(actionNames, new DialogInterface.OnClickListener() {
			public void onClick(DialogInterface dialog, int item) {
				try {
					actions.get(item).execute();
				} catch (Throwable t) {
					UIErrorHandler.showError(t);
				}
			}
		});
		
		AlertDialog alert = builder.create();
		alert.show();
	}
	
	private List<RestoreBackupAction> buildActionsList() {
		List<RestoreBackupAction> actions = new ArrayList<>();
		
		List<Backup> backups = backupManager.getBackups();
		
		for (final Backup backup : backups) {
			actions.add(new RestoreBackupAction(displayDateFormat.format(backup.getDate())) {
				@Override
				public void execute() {
					treeManager.reset();
					scrollCache.clear();
					new PersistenceCommand().loadRootTreeFromBackup(backup);
					new GUICommand().updateItemsList();
					userInfo.showInfo("Database backup loaded: " + backup.getFilename());
				}
			});
		}
		
		return actions;
	}
	
	private CharSequence[] convertToNamesArray(List<RestoreBackupAction> actions) {
		CharSequence[] actionNames = new CharSequence[actions.size()];
		for (int i = 0; i < actions.size(); i++) {
			actionNames[i] = actions.get(i).getName();
		}
		return actionNames;
	}
}

File: app/src/main/java/igrek/todotree/ui/contextmenu/BackupListMenu.kt
package igrek.todotree.ui.contextmenu

import android.app.Activity
import android.app.AlertDialog
import igrek.todotree.info.errorcheck.UiErrorHandler
import igrek.todotree.inject.LazyExtractor
import igrek.todotree.inject.LazyInject
import igrek.todotree.inject.appFactory
import igrek.todotree.intent.PersistenceCommand
import igrek.todotree.service.backup.Backup
import igrek.todotree.service.backup.BackupManager
import java.text.SimpleDateFormat
import java.util.*

class BackupListMenu(
    backupManager: LazyInject<BackupManager> = appFactory.backupManager,
) {
    private val activity: Activity by LazyExtractor(appFactory.activity)

    private val backupManager by LazyExtractor(backupManager)

    private val displayDateFormat: SimpleDateFormat =
        SimpleDateFormat("yyyy-MM-dd, HH:mm:ss", Locale.ENGLISH)

    fun show() {
        val actions = buildActionsList()
        val actionNames = convertToNamesArray(actions)
        val builder = AlertDialog.Builder(activity)
        builder.setTitle("Choose backup")
        builder.setItems(actionNames) { _, item ->
            try {
                actions[item].execute()
            } catch (t: Throwable) {
                UiErrorHandler.handleError(t)
            }
        }
        val alert = builder.create()
        alert.show()
    }

    private fun buildActionsList(): List<RestoreBackupAction> {
        val actions: MutableList<RestoreBackupAction> = ArrayList()
        val backups: List<Backup> = backupManager.getBackups()
        for (backup in backups) {
            actions.add(object : RestoreBackupAction(displayDateFormat.format(backup.date)) {
                override fun execute() {
                    PersistenceCommand().loadRootTreeFromBackup(backup)
                }
            })
        }
        return actions
    }

    private fun convertToNamesArray(actions: List<RestoreBackupAction>): Array<CharSequence?> {
        val actionNames = arrayOfNulls<CharSequence>(actions.size)
        for (i in actions.indices) {
            actionNames[i] = actions[i].name
        }
        return actionNames
    }
}